pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +282,33 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
 		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		volume.Spec.ClaimRef = claimRef
```

Comment:
if anything, I would expect this:
```
volume = ... get from index ...
volumeCopy = copy(volume)
volumeCopy.Spec.ClaimRef = ...
if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeCopy); err != nil {
  return error ...
} else {
  volumeIndex.Update(updatedVolume)
}
```

No rollback needed, no in-memory mutation, but there are several things I am unsure about:
1. Would manually updating the volume index, along with watching volumes result in a double call to `syncVolume`
2. Is there still a race between successfully calling UpdatePersistentVolume and calling `volumeIndex.Update(updatedVolume)` (e.g. could another update event happen, be observed and updated in the volumeIndex, then overwritten by your `Update` call)